libs/architectures/architectures/networks/s4.py

```diff
@@ -0,0 +1,237 @@
+        """
+
+        if lr == 0.0:
+            self.register_buffer(name, tensor)
+        else:
+            self.register_parameter(name, nn.Parameter(tensor))
+
+            optim = {"weight_decay": 0.0}
+            if lr is not None:
+                optim["lr"] = lr
+            setattr(getattr(self, name), "_optim", optim)
+
+
+class S4D(nn.Module):
+    def __init__(
+        self, d_model, d_state=64, dropout=0.0, transposed=True, **kernel_args
+    ):
+        super().__init__()
+
+        self.h = d_model
+        self.n = d_state
+        self.d_output = self.h
+        self.transposed = transposed
+
+        self.D = nn.Parameter(torch.randn(self.h))
+
+        # SSM Kernel
+        self.kernel = S4DKernel(self.h, N=self.n, **kernel_args)
+
+        # Pointwise
+        self.activation = nn.GELU()
+        # dropout_fn = nn.Dropout2d  # NOTE: bugged in PyTorch 1.11
+        dropout_fn = DropoutNd
+        self.dropout = dropout_fn(dropout) if dropout > 0.0 else nn.Identity()
+
+        # position-wise output transform to mix features
+        self.output_linear = nn.Sequential(
+            nn.Conv1d(self.h, 2 * self.h, kernel_size=1),
+            nn.GLU(dim=-2),
+        )
+
+    def forward(
+        self, u, **kwargs
+    ):  # absorbs return_output and transformer src mask
+        """Input and output shape (B, H, L)"""
+        if not self.transposed:
+            u = u.transpose(-1, -2)
+        L = u.size(-1)
+
+        # Compute SSM Kernel
+        k = self.kernel(L=L)  # (H L)
+
+        # Convolution
+        k_f = torch.fft.rfft(k, n=2 * L)  # (H L)
+        u_f = torch.fft.rfft(u, n=2 * L)  # (B H L)
+        y = torch.fft.irfft(u_f * k_f, n=2 * L)[..., :L]  # (B H L)
+
+        # Compute D term in state space equation
+        # Essentially a skip connection
+        y = y + u * self.D.unsqueeze(-1)
+
+        y = self.dropout(self.activation(y))
+        y = self.output_linear(y)
+        if not self.transposed:
+            y = y.transpose(-1, -2)
+        # Return a dummy state to satisfy this repo's interface,
+        # but this can be modified
+        return (
+            y,
+            None,
+        )
+
+
+class S4Model(nn.Module):
+    def __init__(
+        self,
+        d_input,
+        d_output=10,
+        d_model=256,
+        n_layers=4,
+        dropout=0.2,
+        prenorm=False,
+    ):
+        super().__init__()
+
+        self.prenorm = prenorm
+
+        # Linear encoder (d_input = 1 for grayscale and 3 for RGB)
+        self.encoder = nn.Linear(d_input, d_model)
+
+        # Stack S4 layers as residual blocks
+        self.s4_layers = nn.ModuleList()
+        self.norms = nn.ModuleList()
+        self.dropouts = nn.ModuleList()
+        for _ in range(n_layers):
+            self.s4_layers.append(
+                S4D(
+                    d_model,
+                    dropout=dropout,
+                    transposed=True,
+                    # lr=min(0.001, args.lr),
```

Comment:
and then if we do a search I think we will want to mke this an argument somehow

Reply:
Definitely. Might require a little bit of reconfiguring because we don't normally have the learning rate as an architecture parameter, but should be doable

Reply:
I think this lr is supposed to be different from the main learning rate, so we could just expose it as a parameter to the architecture. We could also just move the setting of this learning rate from the architecture to the model

Reply:
Also, would be great to expose the `dt_min`, `dt_max`, and `N` parameters